### DISCUSSION TITLE
Tutorial on firmware decompiling

### DISCUSSION THREAD

**Discussion by DepreTux** [bangcorrupt/hacktribe#89] [Ideas]
Hi!

Is there any chance that you could share the process to decompile the firmware so that the uninitiated like me could start tinkering with it?

I know it's risky, but it may enable a fun learning experience and, who knows, maybe we can help make the hacktribe even more powerful.

Thanks for everything!

**Comment by bangcorrupt** [comment]
Add [AM1802.pspec](https://github.com/bangcorrupt/hacktribe/blob/master/ghidra/AM1802.pspec) to Ghidra along with an entry in [ARM.ldefs](https://github.com/bangcorrupt/hacktribe/blob/master/ghidra/ARM.ldefs).  

Remove the header from the firmware update file and load it into Ghidra at address 0xc0000000.

Ghidra's auto analysis will do a pretty good job identifying the code and decompiling.  

I use rizin for dynamic analysis.  Generally I set a watchpoint on some data or a register, then try to work out what the function accessing it does.

**Reply by DepreTux** [reply]
Thanks a lot again!

**Reply by enorrmann** [reply]
> Remove the header from the firmware update file

Hi ! I'm trying to decompile this, how many bytes is the header ?
new to this
thx !

**Reply by bangcorrupt** [reply]
The header is 256 bytes.  When you import a file into Ghidra you can set an offset, so you wouldn't need to edit the file.

If you want to edit the file, try this:

```bash
dd if=SYSTEM.VSB of=SYSTEM.bin bs=256 skip=1
```

You could check it with hexdump:

```bash
hexdump -vC SYSTEM.VSB | less
```

**Comment by optikfiber** [comment]
Just to chime in on this topic, I'd like to help out if I can but I'm a bit stuck. I've done some reverse engineering years ago but ARM is new to me, so if you could point out eg. where you removed the filter restriction that might be a good starting point to begin understanding how the firmware works better. I can see there's some kind of channel or voice data structure where offset 0x31c holds the filter type, but can't seem to figure out where exactly the original sampler firmware limits you to just the 3 basic types. If I can get a handle on that I can start trying to understand other parts. Probably I'll look into choke groups first since the very limited options annoy me a lot in the original firmware.

**Reply by bangcorrupt** [reply]
If your'e looking for a starting point, the recent MIDI CC hacks are much more organised.  If you diff the patched firmware for commits [364fd9b](https://github.com/bangcorrupt/hacktribe/commit/364fd9b4e413b408060df3e14b21f6e8b09f16b6) and  [58f6560](https://github.com/bangcorrupt/hacktribe/commit/58f65607a9ad7c88972c414f82601605b53de1e9) you should be able to see how I diverted a branch and inserted a function sending MIDI.

**Comment by bukru** [comment]
Ok thanks a lot bangcorrupt ! : )

**Comment by enorrmann** [comment]
when decompiling I encounter several
                    /* WARNING: Bad instruction - Truncating control flow here */
    halt_baddata();

is this OK or am I missing something ?

**Reply by bangcorrupt** [reply]
I see this a lot, I don't know enough about Ghidra to tell you more about it.  I think maybe it's tried to treat data as code, or some opcode is not supported.

**Comment by bangcorrupt** [comment]
#396 @inkapmb

> Tried to dig a bit into synth-revert-SYSTEM.VSB with ghidra

`synth-revert-SYSTEM.VSB` is the factory synth firmware with some bytes changed in the header, the actual firmware has no changes.

> I cant figured out how to use [AM1802.pspec](https://github.com/bangcorrupt/hacktribe/blob/main/ghidra/AM1802.pspec) and [ARM.ldefs](https://github.com/bangcorrupt/hacktribe/blob/main/ghidra/ARM.ldefs).

`AM1802.pspec` goes in `${GHIDRA_INSTALL_DIR}/Ghidra/Processors/ARM/data/languages/`

In `${GHIDRA_INSTALL_DIR}/Ghidra/Processors/ARM/data/languages/ARM.ldefs`, search for the `<language>` element for ARM v5t little endian:

```xml
    <language processor="ARM"
    ...
            id="ARM:LE:32:v5t">
    <description>Generic ARM/Thumb v5 little endian (T-variant)</description>
    ...
    </language>
```

Copy the whole element and change `variant` to `"AM1802"`, `processorspec` to `"AM1802.pspec"`, `id` to `"ARM:LE:32:AM1802"` and update `description`:

```xml
  <language processor="ARM"
            endian="little"
            size="32"
            variant="AM1802"
            version="1.107"
            slafile="ARM5t_le.sla"
            processorspec="AM1802.pspec"
            manualindexfile="../manuals/ARM.idx"
            id="ARM:LE:32:AM1802">
    <description>AM1802 ARM/Thumb v5 little endian (T-variant)</description>
    ...
    </language>
```

On my system, `${GHIDRA_INSTALL_DIR}` is `/opt/ghidra`.  Setting it up this way we have to do it again every time Ghidra is updated.  There is probably a better place to put these definitions so they're not wiped out by updates, but I've never looked it up. 

> add the extension ghidra-blackfin and openned the binary with it

The firmware binary is for the ARM AM1802 CPU, it includes the Blackfin DSP firmware as an array.  Once you have added the pspec and edited ARM.ldefs, import the firmware and set the language to `am1802`.

File > Import File: SYSTEM.VSB
Language: am1802
Options > Base Address: 0xc0000000
Options > File Offset: 0x100

**Reply by inkapmb** [reply]
First thanks a lot for those explanation !

> `synth-revert-SYSTEM.VSB` is the factory synth firmware with some bytes changed in the header, the actual firmware has no changes.

Yes, I would like to see how the synth could work first that why

> Options > Base Address: 0xc0000000
> Options > File Offset: 0x100

I did not notice the options first time, my bad

**Reply by vanana-old-acc** [reply]
So th blackfin code is stored as an array in the 2MiB CPU firmware image? Do you know the address of it?
